File: src/heartlight/synaptic.py
from __future__ import annotations

import math
from collections.abc import Iterable
from dataclasses import dataclass

SYNAPTIC_KERNEL_VERSION = "1.0"


@dataclass(frozen=True, slots=True)
class SynapticConfig:
    learning_rate: float = 0.08
    decay: float = 0.002
    trace_decay: float = 0.9
    min_weight: float = -1.0
    max_weight: float = 1.0

    def validate(self) -> None:
        values = (
            self.learning_rate,
            self.decay,
            self.trace_decay,
            self.min_weight,
            self.max_weight,
        )
        if not all(math.isfinite(value) for value in values):
            raise ValueError("synaptic config values must be finite")
        if self.learning_rate < 0.0:
            raise ValueError("learning_rate must be >= 0")
        if not 0.0 <= self.decay <= 1.0:
            raise ValueError("decay must be between 0 and 1")
        if not 0.0 <= self.trace_decay <= 1.0:
            raise ValueError("trace_decay must be between 0 and 1")
        if self.min_weight > self.max_weight:
            raise ValueError("min_weight must be <= max_weight")


@dataclass(frozen=True, slots=True)
class SynapticState:
    weight: float = 0.0
    eligibility: float = 0.0


@dataclass(frozen=True, slots=True)
class SynapticInput:
    pre: float
    post: float
    reward: float = 1.0


def synaptic_step(
    state: SynapticState,
    sample: SynapticInput,
    config: SynapticConfig | None = None,
) -> SynapticState:
    """Advance one deterministic reward-modulated eligibility-trace synapse.

    e[t+1] = trace_decay * e[t] + pre * post
    w[t+1] = clamp((1-decay) * w[t] + learning_rate * reward * e[t+1])
    """

    cfg = config or SynapticConfig()
    cfg.validate()
    values = (state.weight, state.eligibility, sample.pre, sample.post, sample.reward)
    if not all(math.isfinite(value) for value in values):
        raise ValueError("synaptic state and input values must be finite")

    eligibility = cfg.trace_decay * state.eligibility + sample.pre * sample.post
    weight = (1.0 - cfg.decay) * state.weight + cfg.learning_rate * sample.reward * eligibility
    weight = min(cfg.max_weight, max(cfg.min_weight, weight))
    return SynapticState(weight=weight, eligibility=eligibility)


def synaptic_batch(
    state: SynapticState,
    samples: Iterable[SynapticInput],
    config: SynapticConfig | None = None,
) -> SynapticState:
    cfg = config or SynapticConfig()
    current = state
    for sample in samples:
        current = synaptic_step(current, sample, cfg)
    return current

File: sdk/rust/src/synaptic.rs
use serde::{Deserialize, Serialize};

pub const SYNAPTIC_KERNEL_VERSION: &str = "1.0";

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
pub struct SynapticConfig {
    pub learning_rate: f64,
    pub decay: f64,
    pub trace_decay: f64,
    pub min_weight: f64,
    pub max_weight: f64,
}

impl Default for SynapticConfig {
    fn default() -> Self {
        Self {
            learning_rate: 0.08,
            decay: 0.002,
            trace_decay: 0.9,
            min_weight: -1.0,
            max_weight: 1.0,
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
pub struct SynapticState {
    pub weight: f64,
    pub eligibility: f64,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
pub struct SynapticInput {
    pub pre: f64,
    pub post: f64,
    pub reward: f64,
}

pub fn synaptic_step(state: SynapticState, sample: SynapticInput, config: SynapticConfig) -> SynapticState {
    assert!(config.learning_rate.is_finite() && config.learning_rate >= 0.0);
    assert!(config.decay.is_finite() && (0.0..=1.0).contains(&config.decay));
    assert!(config.trace_decay.is_finite() && (0.0..=1.0).contains(&config.trace_decay));
    assert!(config.min_weight.is_finite() && config.max_weight.is_finite());
    assert!(config.min_weight <= config.max_weight);
    assert!(state.weight.is_finite() && state.eligibility.is_finite());
    assert!(sample.pre.is_finite() && sample.post.is_finite() && sample.reward.is_finite());

    let eligibility = config.trace_decay * state.eligibility + sample.pre * sample.post;
    let weight = ((1.0 - config.decay) * state.weight
        + config.learning_rate * sample.reward * eligibility)
        .clamp(config.min_weight, config.max_weight);
    SynapticState { weight, eligibility }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn conformance_v1() {
        let config = SynapticConfig::default();
        let mut state = SynapticState { weight: 0.2, eligibility: 0.0 };
        let steps = [
            (SynapticInput { pre: 0.5, post: 0.25, reward: 1.0 }, 0.2096, 0.125),
            (SynapticInput { pre: 1.0, post: -0.5, reward: 0.4 }, 0.1967808, -0.3875),
            (SynapticInput { pre: 0.2, post: 0.8, reward: -0.25 }, 0.2001622384, -0.18875),
        ];
        for (sample, expected_weight, expected_eligibility) in steps {
            state = synaptic_step(state, sample, config);
            assert!((state.weight - expected_weight).abs() < 1e-12);
            assert!((state.eligibility - expected_eligibility).abs() < 1e-12);
        }
    }
}
